File: services/caregivers.js
import { db } from '../db/client.js';
import { caregivers, seniors } from '../db/schema.js';
import { eq, and } from 'drizzle-orm';

export const caregiverService = {
  // Link a Clerk user to a senior (creates caregiver assignment, skips if already linked)
  async linkUserToSenior(clerkUserId, seniorId, role = 'caregiver') {
    // Check if already linked
    const existing = await this.getAssignment(clerkUserId, seniorId);
    if (existing) {
      console.log(`[Caregiver] User ${clerkUserId} already linked to senior ${seniorId}, skipping`);
      return existing;
    }

    const [assignment] = await db.insert(caregivers).values({
      clerkUserId,
      seniorId,
      role,
    }).returning();

    console.log(`[Caregiver] Linked user ${clerkUserId} to senior ${seniorId} as ${role}`);
    return assignment;
  },

  // Get all seniors accessible by a Clerk user
  async getSeniorsForUser(clerkUserId) {
    const assignments = await db.select({
      assignment: caregivers,
      senior: seniors,
    })
      .from(caregivers)
      .innerJoin(seniors, eq(caregivers.seniorId, seniors.id))
      .where(and(
        eq(caregivers.clerkUserId, clerkUserId),
        eq(seniors.isActive, true)
      ));

    return assignments.map(a => ({
      ...a.senior,
      role: a.assignment.role,
    }));
  },

  // Check if a Clerk user can access a senior
  async canAccessSenior(clerkUserId, seniorId) {
    const [assignment] = await db.select()
      .from(caregivers)
      .where(and(
        eq(caregivers.clerkUserId, clerkUserId),
        eq(caregivers.seniorId, seniorId)
      ))
      .limit(1);
    return !!assignment;
  },

  // Get all Clerk users who can access a senior
  async getUsersForSenior(seniorId) {
    const assignments = await db.select()
      .from(caregivers)
      .where(eq(caregivers.seniorId, seniorId));

    return assignments.map(a => ({
      clerkUserId: a.clerkUserId,
      role: a.role,
    }));
  },

  // Remove a user's access to a senior
  async unlinkUserFromSenior(clerkUserId, seniorId) {
    const result = await db.delete(caregivers)
      .where(and(
        eq(caregivers.clerkUserId, clerkUserId),
        eq(caregivers.seniorId, seniorId)
      ))
      .returning();

    return result.length > 0;
  },

  // Get assignment details
  async getAssignment(clerkUserId, seniorId) {
    const [assignment] = await db.select()
      .from(caregivers)
      .where(and(
        eq(caregivers.clerkUserId, clerkUserId),
        eq(caregivers.seniorId, seniorId)
      ));
    return assignment || null;
  },
};

File: pipecat/services/caregivers.py
"""Caregiver service.

Port of services/caregivers.js — manages caregiver-senior relationships via Clerk user IDs.
"""

from __future__ import annotations

from loguru import logger
from db import query_one, query_many, execute


async def link_user_to_senior(
    clerk_user_id: str, senior_id: str, role: str = "caregiver"
) -> dict:
    """Link a Clerk user to a senior (creates caregiver assignment)."""
    row = await query_one(
        """INSERT INTO caregivers (clerk_user_id, senior_id, role)
           VALUES ($1, $2, $3)
           RETURNING *""",
        clerk_user_id,
        senior_id,
        role,
    )
    logger.info("Linked user {uid} to senior {sid} as {role}", uid=clerk_user_id, sid=senior_id, role=role)
    return row


async def get_seniors_for_user(clerk_user_id: str) -> list[dict]:
    """Get all seniors accessible by a Clerk user."""
    rows = await query_many(
        """SELECT s.*, c.role
           FROM caregivers c
           INNER JOIN seniors s ON c.senior_id = s.id
           WHERE c.clerk_user_id = $1 AND s.is_active = true""",
        clerk_user_id,
    )
    return rows


async def can_access_senior(clerk_user_id: str, senior_id: str) -> bool:
    """Check if a Clerk user can access a senior."""
    row = await query_one(
        """SELECT id FROM caregivers
           WHERE clerk_user_id = $1 AND senior_id = $2
           LIMIT 1""",
        clerk_user_id,
        senior_id,
    )
    return row is not None


async def get_users_for_senior(senior_id: str) -> list[dict]:
    """Get all Clerk users who can access a senior."""
    rows = await query_many(
        "SELECT clerk_user_id, role FROM caregivers WHERE senior_id = $1",
        senior_id,
    )
    return rows


async def unlink_user_from_senior(clerk_user_id: str, senior_id: str) -> bool:
    """Remove a user's access to a senior. Returns True if a row was deleted."""
    result = await execute(
        "DELETE FROM caregivers WHERE clerk_user_id = $1 AND senior_id = $2",
        clerk_user_id,
        senior_id,
    )
    # asyncpg execute returns 'DELETE N'
    return not result.endswith("0")


async def get_assignment(clerk_user_id: str, senior_id: str) -> dict | None:
    """Get assignment details for a specific user-senior pair."""
    return await query_one(
        "SELECT * FROM caregivers WHERE clerk_user_id = $1 AND senior_id = $2",
        clerk_user_id,
        senior_id,
    )
